services/file_util.cpp

```diff
@@ -69,50 +71,49 @@ string get_relative_path(const string &to, const string &from) {
 }
 
 /**
- * Returns a string without '..' and '.'
+ * Return a canonicalized absolute path, with symlinks resolved (if possible).
  *
- * Preconditions:  path must be an absolute path
- * Postconditions: if path is empty or not an absolute path, return original
- *                 path, otherwise, return path after resolving '..' and '.'
+ * If part of the path does not currently exist, make a reasonable guess by
+ * ignoring symlinks.
  */
-string get_canonicalized_path(const string &path) {
-    if (path.empty() || path[0] != '/') {
+string get_abspath(const string &path)
+{
+    if (path.empty()) {
         return path;
     }
 
-    vector<string> parts = split(path, '/');
-    vector<string> canonicalized_path;
-
-    vector<string>::const_iterator parts_it = parts.begin(),
-                                   parts_end = parts.end();
-
-    while (parts_it != parts_end) {
-        if (*parts_it == ".." && !canonicalized_path.empty()) {
-            canonicalized_path.pop_back();
-        }
-        else if (*parts_it != "." && *parts_it != "..") {
-            canonicalized_path.push_back(*parts_it);
-        }
-
-        ++parts_it;
+    char buf[PATH_MAX];
+    buf[PATH_MAX - 1] = '\0';
+    if (realpath(path.c_str(), buf) != NULL) {
+        return string(buf);
     }
 
-    vector<string>::const_iterator path_it = canonicalized_path.begin(),
-                                   path_end = canonicalized_path.end();
+    if (errno != ENOENT) {
+        log_error() << "Unable to canonicalize path: "
+                    << strerror(errno) << endl;
+        return path;
+    }
 
-    string output;
-    output.reserve(path.size());
-    output += "/";
-    while (path_it != path_end) {
-        output += *path_it;
+    log_info() << "Path doesn't exist yet, guessing the canonical form."
+               << endl;
 
-        ++path_it;
-        if (path_it != path_end) {
-            output += "/";
+    vector<string> components = split(path, '/');
+    deque<string> stack;
```

Comment:
`std::vector<string>` is probably faster here, actually. In fact:

```cpp
vector<string> stack(components.size());
```
Would allocate it to be as big as it'll ever need to be so there will never be a reallocation.

Reply:
We're modelling a directory stack, switching to std::vector would require the stack-ness of this to be handled manually, which would introduce some complexity.  I have a feeling that there would not be a huge performance difference in practice here- in my experience directory depths for compilation are typically under say 20.

I picked std::deque because I want to treat it like a stack during processing, but then like a queue when building the output.  It shouldn't be difficult to use std::queue instead, then build the resulting string backwards.  Does that sound efficient enough (the code should be simpler than using std::vector)?

Reply:
You mean http://en.cppreference.com/w/cpp/container/vector/pop_back doesn't do what you need? I think `std::vector` is fine here.

Reply:
Pushed a new version which scans through the components vector twice, without using std::deque or similar.